alacritty_terminal/src/grid/storage.rs

```diff
@@ -312,6 +336,75 @@ mod test {
     use crate::grid::storage::Storage;
     use crate::index::{Column, Line};
 
+    #[test]
+    fn with_capacity() {
+        let storage = Storage::with_capacity(Line(3),
+                                             Row::new(Column(0), &' '));
+
+        assert_eq!(storage.inner.len(), 3);
+        assert_eq!(storage.len, 3);
+        assert_eq!(storage.zero, 0);
+        assert_eq!(storage.visible_lines, Line(2));
+    }
+
+    #[test]
+    fn indexing() {
+        let mut storage = Storage::with_capacity(Line(3),
+                                                 Row::new(Column(0), &' '));
+
+        storage[0] = Row::new(Column(1), &'0');
+        storage[1] = Row::new(Column(1), &'1');
+        storage[2] = Row::new(Column(1), &'2');
+
+        assert_eq!(storage[0], Row::new(Column(1), &'0'));
+        assert_eq!(storage[1], Row::new(Column(1), &'1'));
+        assert_eq!(storage[2], Row::new(Column(1), &'2'));
+
+        storage.zero += 1;
+
+        assert_eq!(storage[0], Row::new(Column(1), &'1'));
+        assert_eq!(storage[1], Row::new(Column(1), &'2'));
+        assert_eq!(storage[2], Row::new(Column(1), &'0'));
+    }
+
+    #[test]
+    #[should_panic]
+    fn indexing_above_len() {
+        let mut storage = Storage::with_capacity(Line(3),
+                                                 Row::new(Column(0), &' '));
+        storage.shrink_lines(2);
+        &storage[1];
+    }
+
+    #[test]
+    #[should_panic]
+    fn indexing_above_inner_len() {
+        let storage = Storage::with_capacity(Line(0),
+                                             Row::new(Column(0), &' '));
+        &storage[1];
+    }
+
+    #[test]
+    fn rotate() {
+        let mut storage = Storage::with_capacity(Line(3),
+                                                 Row::new(Column(0), &' '));
+        storage.rotate(2);
+        assert_eq!(storage.zero, 2);
+        storage.shrink_lines(2);
+        // TODO: Shouldn't `shrink_lines` update the value of `zero`?
+        assert_eq!(storage.len, 1);
+        assert_eq!(storage.zero, 0);
```

Comment:
@chrisduerr I'm not 100% sure what I expect here. Thoughts?

(This is probably unrelated to me issues with the PR atm I fear however)

Reply:
Why would zero move? It just reduces the number of elements. All that requires to do is reduce the `len` and we should be good, the rest is then automatically treated as excess buffer.

Zero is reset though when more than a certain number of lines have been removed, to prevent excessive memory consumption.

Reply:
That relies on every use of `zero` correctly handling the case where it's outside the inner length. But this does seem to be the case, so ignore this comment.